evm_arithmetization/src/generation/state.rs

```diff
@@ -1,26 +1,217 @@
 use std::collections::HashMap;
 
+use anyhow::bail;
 use ethereum_types::{Address, BigEndianHash, H160, H256, U256};
 use keccak_hash::keccak;
+use log::log_enabled;
 use plonky2::field::types::Field;
 
 use super::mpt::{load_all_mpts, TrieRootPtrs};
 use super::TrieInputs;
 use crate::cpu::kernel::aggregator::KERNEL;
 use crate::cpu::kernel::constants::context_metadata::ContextMetadata;
+use crate::cpu::membus::NUM_GP_CHANNELS;
+use crate::cpu::stack::MAX_USER_STACK_SIZE;
 use crate::generation::rlp::all_rlp_prover_inputs_reversed;
+use crate::generation::CpuColumnsView;
 use crate::generation::GenerationInputs;
 use crate::memory::segments::Segment;
 use crate::util::u256_to_usize;
 use crate::witness::errors::ProgramError;
-use crate::witness::memory::{MemoryAddress, MemoryState};
+use crate::witness::memory::MemoryChannel::GeneralPurpose;
+use crate::witness::memory::MemoryOpKind;
+use crate::witness::memory::{MemoryAddress, MemoryOp, MemoryState};
+use crate::witness::operation::{generate_exception, Operation};
 use crate::witness::state::RegistersState;
 use crate::witness::traces::{TraceCheckpoint, Traces};
-use crate::witness::util::stack_peek;
+use crate::witness::transition::{
+    decode, fill_op_flag, get_op_special_length, log_kernel_instruction, might_overflow_op,
+    read_code_memory, Transition,
+};
+use crate::witness::util::{
+    fill_channel_with_value, mem_read_gp_with_log_and_fill, stack_peek, stack_pop_with_log_and_fill,
+};
 
-pub(crate) struct GenerationStateCheckpoint {
-    pub(crate) registers: RegistersState,
-    pub(crate) traces: TraceCheckpoint,
+/// A State is either an `Interpreter` (used for tests and jumpdest analysis) or
+/// a `GenerationState`.
+pub(crate) trait State<F: Field> {
+    /// Returns a `State`'s `Checkpoint`.
+    fn checkpoint(&mut self) -> GenerationStateCheckpoint;
+
+    /// Increments the `gas_used` register by a value `n`.
+    fn incr_gas(&mut self, n: u64);
+
+    /// Increments the `program_counter` register by a value `n`.
+    fn incr_pc(&mut self, n: usize);
+
+    /// Returns a `State`'s registers.
+    fn get_registers(&self) -> RegistersState;
+
+    /// Returns a `State`'s mutable registers.
+    fn get_mut_registers(&mut self) -> &mut RegistersState;
+
+    /// Returns the value stored at address `address` in a `State`.
+    fn get_from_memory(&mut self, address: MemoryAddress) -> U256;
+
+    /// Returns a mutable `GenerationState` from a `State`.
+    fn get_mut_generation_state(&mut self) -> &mut GenerationState<F>;
+
+    /// Returns true if a `State` is a `GenerationState` and false otherwise.
+    fn is_generation_state(&mut self) -> bool;
```

Comment:
It would be nice if we can get rid of this function as we must always statically know if `self` is a `GenerationState` or not.